File: native/ios/Sources/AccountRegistration.swift
import Combine
import Foundation

@MainActor
protocol RegistrationService {
    func register(username: String, password: String, name: String) async throws -> Bool
    func confirmRegistration(username: String, code: String) async throws -> Bool
    func resendRegistration(username: String) async throws
}

enum RegistrationStep: Hashable { case closed, details, confirm, complete }
struct RegistrationState: Equatable {
    var step: RegistrationStep = .closed
    var accepted = false
    var busy = false
    var failed = false
    var codeRequested = false
}

@MainActor
final class AccountRegistration: ObservableObject {
    @Published private(set) var state = RegistrationState()
    private let service: RegistrationService?
    private var username: String?
    private var operation: Task<Void, Never>?
    private var generation = 0
    init(service: RegistrationService?) { self.service = service }
    func close() {
        generation += 1; operation?.cancel(); operation = nil; username = nil
        state = RegistrationState()
    }
    func open() { close(); if service != nil { state = RegistrationState(step: .details) } }
    func accept(_ value: Bool) {
        if state.step == .details && !state.busy { state.accepted = value }
    }
    func resumeConfirmation(_ value: String) {
        close()
        let value = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if service != nil && !value.isEmpty { username = value; state = RegistrationState(step: .confirm) }
    }
    func register(email: String, password: String, name: String) {
        let subject = email.trimmingCharacters(in: .whitespacesAndNewlines)
        let name = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let service, state.step == .details, !state.busy, state.accepted,
              !subject.isEmpty, !name.isEmpty, !password.isEmpty else { return }
        username = subject
        run { [weak self] in
            let complete = try await service.register(username: subject, password: password, name: name)
            try Task.checkCancellation()
            if complete { self?.username = nil }
            self?.state = RegistrationState(step: complete ? .complete : .confirm)
        }
    }
    func confirm(_ code: String) {
        let code = code.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let service, let username, state.step == .confirm, !state.busy, !code.isEmpty else { return }
        run { [weak self] in
            let complete = try await service.confirmRegistration(username: username, code: code)
            try Task.checkCancellation()
            if complete { self?.username = nil; self?.state = RegistrationState(step: .complete) }
            else { self?.state.failed = true }
        }
    }
    func resend() {
        guard let service, let username, state.step == .confirm, !state.busy else { return }
        run { [weak self] in
            try await service.resendRegistration(username: username)
            try Task.checkCancellation()
            self?.state.codeRequested = true
        }
    }
    private func run(_ block: @escaping @MainActor () async throws -> Void) {
        generation += 1
        let current = generation
        state.busy = true; state.failed = false; state.codeRequested = false
        operation = Task { [weak self] in
            do { try await block() }
            catch is CancellationError { return }
            catch {
                guard let self, current == generation, !Task.isCancelled else { return }
                state.failed = true
            }
            if let self, current == generation { state.busy = false }
        }
    }
}

File: native/android/app/src/main/java/com/mediconnect/nativeapp/AccountRegistration.kt
package com.mediconnect.nativeapp

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

interface RegistrationService {
    suspend fun register(username: String, password: String, name: String): Boolean
    suspend fun confirmRegistration(username: String, code: String): Boolean
    suspend fun resendRegistration(username: String)
}

enum class RegistrationStep { CLOSED, DETAILS, CONFIRM, COMPLETE }
data class RegistrationState(val step: RegistrationStep = RegistrationStep.CLOSED,
    val accepted: Boolean = false, val busy: Boolean = false, val failed: Boolean = false, val codeRequested: Boolean = false)

class AccountRegistration(private val service: RegistrationService?, private val scope: CoroutineScope) {
    private val mutable = MutableStateFlow(RegistrationState())
    val state = mutable.asStateFlow()
    private var username: String? = null
    private var operation: Job? = null
    private var generation = 0L

    fun close() {
        generation++; operation?.cancel(); operation = null; username = null
        mutable.value = RegistrationState()
    }
    fun open() { close(); if (service != null) mutable.value = RegistrationState(RegistrationStep.DETAILS) }
    fun accept(value: Boolean) {
        if (mutable.value.step == RegistrationStep.DETAILS && !mutable.value.busy) {
            mutable.value = mutable.value.copy(accepted = value)
        }
    }
    fun resumeConfirmation(value: String) {
        close()
        if (service != null && value.isNotBlank()) {
            username = value.trim(); mutable.value = RegistrationState(RegistrationStep.CONFIRM)
        }
    }
    fun register(email: String, password: String, name: String) {
        val service = service ?: return
        val current = mutable.value
        if (current.step != RegistrationStep.DETAILS || current.busy || !current.accepted ||
            email.isBlank() || name.isBlank() || password.isEmpty()) return
        val subject = email.trim(); username = subject
        run {
            val complete = service.register(subject, password, name.trim())
            currentCoroutineContext().ensureActive()
            if (complete) username = null
            mutable.value = RegistrationState(if (complete) RegistrationStep.COMPLETE else RegistrationStep.CONFIRM)
        }
    }
    fun confirm(code: String) {
        val service = service ?: return
        val subject = username ?: return
        if (mutable.value.step != RegistrationStep.CONFIRM || mutable.value.busy || code.isBlank()) return
        run {
            val complete = service.confirmRegistration(subject, code.trim())
            currentCoroutineContext().ensureActive()
            if (complete) { username = null; mutable.value = RegistrationState(RegistrationStep.COMPLETE) }
            else mutable.value = mutable.value.copy(failed = true)
        }
    }
    fun resend() {
        val service = service ?: return
        val subject = username ?: return
        if (mutable.value.step != RegistrationStep.CONFIRM || mutable.value.busy) return
        run {
            service.resendRegistration(subject)
            currentCoroutineContext().ensureActive()
            mutable.value = mutable.value.copy(codeRequested = true)
        }
    }
    private fun run(block: suspend () -> Unit) {
        val current = ++generation
        mutable.value = mutable.value.copy(busy = true, failed = false, codeRequested = false)
        operation = scope.launch {
            try { block() }
            catch (cancelled: CancellationException) { throw cancelled }
            catch (_: Exception) { if (current == generation) mutable.value = mutable.value.copy(failed = true) }
            finally { if (current == generation) mutable.value = mutable.value.copy(busy = false) }
        }
    }
}
